rfcs/0014-abi-errors.md

```diff
@@ -0,0 +1,456 @@
+conditions, so we need to improve this.
+
+We want irrecoverable errors to carry user defined error messages and source
+information about where the error was invoked. We want those error types to be
+something that can also be used for recoverable errors.
+
+To this end we want to introduce a mechanism that makes error codes an
+understandable part of the ABI, and language features that allow the user to
+define error messages, error types and invoke those errors with the appropriate
+information being generated.
+
+We also want recoverable errors to connect to this mechanism so that all error
+handling in a given Sway program can go through one or a small set of user
+defined types and error strings don't affect the final bytecode size.
+
+# Guide-level explanation
+
+[guide-level-explanation]: #guide-level-explanation
+
+## Error types
+
+Error types are special enums that allow developers to signal that each variant
+is associated with an error message.
+
+An error type implements the `Error` trait, but it is not recommended that you
+do so manually.
+
+Instead, to create a new error type, use the following syntax and annotations on
+an enum:
+
+```sway
+#[error_type]
+enum MyError {
+  #[error("error A")]
+  A: (),
+  #[error("error B")]
+  B: (u64, u8),
+}
+```
+
+This will automatically implement the `Error` marker trait for this enum
+and let the compiler know what error messages correspond to what variants of
+the enum.
+
+It will also populate the ABI specification file with error messages in the
+metadata of the enum type. Allowing users to check the meaning of an error
+without the need to store and decode error messages on chain.
+
+For instance, the following contract method may return a `MyError` variant which
+will be compact but can still be associated with an error message:
+
+```sway
+fn do_something(self) -> Result<(), MyError> {
+  // ...
+  return Err(MyError::A); // user will know that this means "error A"
+  // ...
+}
+```
+
+## `panic`
+
+When encountering an irrecoverable error in a sway program, it is customary to
+revert and produce an informative error code.
+
+The recommended way of doing this is to use the `panic` intrinsic.
+
+`panic` is a compiler intrinsic that will
+produce a unique revert code for each of its invocations and populate the ABI
+specification file with a list of such codes and corresponding information about
+where it was invoked and with what kind of arguments.
+
+If all you want is an error message, you can panic with a string literal:
+
+```sway
+// ...
+// some error happened
+panic("some error happened");
+```
+
+But if you want more functionality, you should use an error type:
+
+```sway
+panic(MyError::A);
+// ...
+panic(MyError::B((1024, 42)));
+```
+
+Panicking with an error type will do two things: log the error so that a
+corresponding log receipt is produced, and revert with an error code that
+corresponds to the invocation location.
```

Comment:
There's no real way we can do a backtrace at compile time since you don't know where you're going to be called from.

So yeah it's more about tracking user errors and having some kind of consistent error message than providing full causation information.

I'm open to suggestions on how we could do something like this in a later change, however.

Reply:
@tritao This is an interesting approach and I see it worth exploring further. My first feeling is that it could work and be very helpful in the debug mode, during the local development. It would give a development experience close to that in "regular" programming. Stacktraces in case of panics etc.

But I see the cost as too high for deployed code.

> There is a very small overhead with this.

I have serious doubts that this would be the case. E.g., right now, the cost of just calculating return addresses during function calls took 2k of 11k of bytecode in a predicate we analyzed. We don't have a notion of stack frames at all on the VM side, function calls are only jumps etc. etc. Manually building that information in memory and keeping track of it at runtime, I cannot imagine that it can be done without exploding the bytecode size. Unless we get specialized VM opcodes. Then of course there is the gas cost attached to it.

But combining the two approaches, this I can definitely imagine:
- The approach defined in this RFC is used for deployed production code. The combination of the compiler generated info and well designed API is indeed lacking the stack trace but still gives rich information for easier analysis of production errors.
- The approach in the comment is used for local development and debugging.

Reply:
I agree that we'll want a full stack trace for the debugger, and that the stack trace should come from the VM.

Reply:
I think we all agree that stack traces would be nice but that we're not to burden the resulting final binaries with overhead.

This probably is beyond this particular RFC, I just want to make sure that we don't block ourselves from that by implementing this.